File: StriverA2Z/JavaScript&TypeScript/src/BinaryTrees/BinaryTreeOperations.ts
export class BTNode {
	data: any;
	left: BTNode | null;
	right: BTNode | null;

	constructor(data: any) {
		this.data = data;
		this.left = null;
		this.right = null;
	}
}

export function preorderTraversal(node: BTNode) {
	if (node === null) return;
	console.log(node.data);
	preorderTraversal(node.left);
	preorderTraversal(node.right);
}

export function inorderTraversal(node: BTNode) {
	if (node === null) return;
	inorderTraversal(node.left);
	console.log(node.data);
	inorderTraversal(node.right);
}

export function postorderTraversal(node: BTNode) {
	if (node === null) return;
	postorderTraversal(node.left);
	postorderTraversal(node.right);
	console.log(node.data);
}

export function arrayToBTLevelOrder(arr: any[]) {
	if (arr.length === null) return null;
	let root = new BTNode(arr[0]);
	const q = [];
	q.push(root);
	let i = 0;
	let n = (Math.floor(arr.length / 2));
	while (q.length !== 0 && i < n) {
		let node = q.shift();
		let left = new BTNode(arr[(2 * i) + 1]);
		node.left = left;
		q.push(left);

		if ((2*i+2) !== arr.length) {
			let right = new BTNode(arr[(2 * i) + 2]);
			node.right = right;
			q.push(right);
		}
		i++;
	}
	return root;
}

export function levelOrder(root: BTNode) {
	if (root === null) return;
	const q = [];
	q.push(root);
	console.log(root.data);
	while (q.length !== 0) {
		for (let i = 0; i < q.length; i++) {
			let node = q.shift();
			if (node.left) console.log(node.left.data);
			if (node.right) console.log(node.right.data);
			if (node.left) q.push(node.left);
			if (node.right) q.push(node.right);
		}
	}
}

export function preorderTraversalIterative(root: BTNode) {
	if (root === null) return;
	const st = [];
	st.push(root);
	while (st.length !== 0) {
		let node = st.pop();
		console.log(node.data);
		if (node.right) st.push(node.right);
		if (node.left) st.push(node.left);
	}
}

export function inorderTraversalIterative(root: BTNode) {
	if (root === null) return;
	const st = [];
	let node = root;
	while (true) {
		if (node) {
			st.push(node);
			node = node.left;
		} else {
			if (st.length === 0) break;
			node = st.pop();
			console.log(node.data);
			node = node.right;
		}
	}
}

export function postorderTraversalIterative2Stack(root: BTNode) {
	if (root === null) return;
	const st1 = [];
	const st2 = [];
	st1.push(root);
	while (st1.length !== 0) {
		let node  = st1.pop();
		st2.push(node);
		if (node.left) st1.push(node.left);
		if (node.right) st1.push(node.right);
	}
	while (st2.length !== 0) {
		console.log(st2.pop().data);
	}
}

export function postorderTraversalIterative1Stack(root: BTNode) {
	if (root === null) return;
	const st = [];
	let curr = root;
	while (curr || st.length !== 0) {
		if (curr) {
			st.push(curr);
			curr = curr.left;
		} else {
			let temp = st[st.length - 1].right;
			if (temp === null) {
				temp = st.pop();
				console.log(temp.data);
				while (st.length !== 0 && temp === st[st.length - 1].right) {
					temp = st.pop();
					console.log(temp.data)
				}
			} else {
				curr = temp;
			}
		}
	}
}

export function preInPostTraversal(root: BTNode) {
	if (root === null) return;
	const arr = []
	arr.push([root, 1]);
	const pre = [];
	const post = [];
	const inorder = [];
	while (arr.length !== 0) {
		let [node, num] = arr.pop();
		if (num === 1) {
			pre.push(node.data);
			num++;
			arr.push([node,num]);
			if (node.left) arr.push([node.left, 1]);
		} else if (num === 2) {
			inorder.push(node.data);
			num++;
			arr.push([node, num]);
			if (node.right) arr.push([node.right, 1]);
		} else {
			post.push(node.data);
		}
	}

	console.log(pre)
	console.log(inorder)
	console.log(post)
}

export function arrayToBT(arr: any[], index = 0) {
	// Base case: If the current index is out of bounds or the element is null, return null
	if (index >= arr.length || arr[index] === null) {
		return null;
	}

	// Create a new TreeNode with the current element
	const root = new BTNode(arr[index]);

	// Recursively build the left and right subtrees
	root.left = arrayToBT(arr, 2 * index + 1);
	root.right = arrayToBT(arr, 2 * index + 2);

	return root;
}

let root = arrayToBTLevelOrder([1, 2, 3, 4, 5, 6, 7]);

// preorderTraversal(root);
// console.log();
// postorderTraversal(root);
// console.log();
// inorderTraversal(root);
// console.log();
// levelOrder(root);
// console.log();
// preorderTraversalIterative(root);
// console.log();
// inorderTraversalIterative(root);
// console.log();
// postorderTraversalIterative2Stack(root);
// console.log();
// postorderTraversalIterative1Stack(root);

// preInPostTraversal(root);

File: StriverA2Z/Java/BinaryTrees/BinaryTreeOperations.java
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.Queue;
import java.util.Stack;

public class BinaryTreeOperations {
    public static void main(String[] args) {
        Node root = arrayToBTLevelOrder(new Integer[]{1, 2, 3, 4, 5, 6, 7});

        preorderTraversal(root);
        System.out.println();

        inorderTraversal(root);
        System.out.println();

        postorderTraversal(root);
        System.out.println();

        levelOrder(root);
        System.out.println();

        preorderTraversalIterative(root);
        System.out.println();

        inorderTraversalIterative(root);
        System.out.println();

        postorderTraversalIterative2Stack(root);
        System.out.println();

        postorderTraversalIterative1Stack(root);
        System.out.println();

        preInPostTraversal(root);
    }

    public static void preorderTraversal(Node node) {
        if (node == null) return;
        System.out.print(node.data + " ");
        preorderTraversal(node.left);
        preorderTraversal(node.right);
    }

    public static void inorderTraversal(Node node) {
        if (node == null) return;
        inorderTraversal(node.left);
        System.out.print(node.data + " ");
        inorderTraversal(node.right);
    }

    public static void postorderTraversal(Node node) {
        if (node == null) return;
        postorderTraversal(node.left);
        postorderTraversal(node.right);
        System.out.print(node.data + " ");
    }

    public static Node arrayToBTLevelOrder(Integer[] arr) {
        if (arr.length == 0 || arr[0] == null) return null;

        Node root = new Node(arr[0]);
        int i = 0;

        Queue<Node> q = new LinkedList<>();
        q.add(root);

        while (!q.isEmpty() && i < (arr.length / 2)) {
            Node t = q.poll();

            Node left = new Node(arr[2 * i + 1]);
            t.left = left;
            q.offer(left);

            if ((2 * i + 2) != arr.length) {
                Node right = new Node(arr[2 * i + 2]);
                t.right = right;
                q.offer(right);
            }
            i++;
        }
        return root;
    }

    public static void levelOrder(Node root) {
        if (root == null) return;
        Queue<Node> q = new LinkedList<>();
        q.offer(root);
        System.out.print(root.data + " ");
        while (!q.isEmpty()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                Node node = q.peek();
                q.poll();
                if (node.left != null) System.out.print(node.left.data + " ");
                if (node.right != null) System.out.print(node.right.data + " ");
                if (node.left != null) q.offer(node.left);
                if (node.right != null) q.offer(node.right);
            }
        }
    }

    public static void preorderTraversalIterative(Node root) {
        if (root == null) return;
        Stack<Node> st = new Stack<>();
        st.push(root);
        while (!st.isEmpty()) {
            Node node = st.pop();
            System.out.print(node.data + " ");
            if (node.right != null) st.push(node.right);
            if (node.left != null) st.push(node.left);
        }
    }

    public static void inorderTraversalIterative(Node root) {
        if (root == null) return;
        Stack<Node> st = new Stack<>();
        Node node = root;
        while (true) {
            if (node != null) {
                st.push(node);
                node = node.left;
            } else {
                if (st.isEmpty()) break;
                node = st.pop();
                System.out.print(node.data + " ");
                node = node.right;
            }
        }
    }

    public static void postorderTraversalIterative2Stack(Node root) {
        if (root == null) return;
        Stack<Node> st1 = new Stack<>();
        Stack<Node> st2 = new Stack<>();
        st1.push(root);
        while (!st1.isEmpty()) {
            Node node = st1.pop();
            st2.push(node);
            if (node.left != null) st1.push(node.left);
            if (node.right != null) st1.push(node.right);
        }
        while (!st2.isEmpty()) {
            System.out.print(st2.peek().data + " ");
            st2.pop();
        }
    }

    public static void postorderTraversalIterative1Stack(Node root) {
        if (root == null) return;
        Stack<Node> st = new Stack<>();
        Node curr = root;
        while (curr != null || !st.isEmpty()) {
            if (curr != null) {
                st.push(curr);
                curr = curr.left;
            } else {
                Node temp = st.peek().right;
                if (temp == null) {
                    temp = st.pop();
                    System.out.print(temp.data + " ");
                    while (!st.isEmpty() && temp == st.peek().right) {
                        temp = st.pop();
                        System.out.print(temp.data + " ");
                    }
                } else {
                    curr = temp;
                }
            }
        }
    }

    public static void preInPostTraversal(Node root) {
        if (root == null) return;
        Stack<Pair> st = new Stack<>();
        ArrayList<Integer> pre = new ArrayList<>();
        ArrayList<Integer> in = new ArrayList<>();
        ArrayList<Integer> post = new ArrayList<>();
        st.push(new Pair(root, 1));
        while (!st.isEmpty()) {
            Pair it = st.pop();
            if (it.num == 1) {
                pre.add(it.node.data);
                it.num++;
                st.push(it);
                if (it.node.left != null) st.push(new Pair(it.node.left, 1));
            } else if (it.num == 2) {
                in.add(it.node.data);
                it.num++;
                st.push(it);
                if (it.node.right != null) st.push(new Pair(it.node.right, 1));
            } else {
                post.add(it.node.data);
            }
        }

        for (int i : pre) {
            System.out.print(i + " ");
        }
        System.out.println();

        for (int i : in) {
            System.out.print(i + " ");
        }
        System.out.println();

        for (int i : post) {
            System.out.print(i + " ");
        }
        System.out.println();
    }

    public static Node arrayToBT(Integer[] arr, int index) {
        if (index >= arr.length || (arr[index] == -1)) return null;
        Node root = new Node(arr[index]);
        root.left = arrayToBT(arr, 2 * index + 1);
        root.right = arrayToBT(arr, 2 * index + 2);
        return root;
    }
}
